File: OS/code/readerwriter.cpp
#include <stdio.h>
#include <pthread.h>
#include <semaphore.h>
#include <unistd.h> // For sleep()
#include <cstdint> 

sem_t mutex; // Mutex for reader count access
sem_t db;    // Semaphore for database access
int readercount = 0;

void *reader(void *p);
void *writer(void *p);

int main() {
    sem_init(&mutex, 0, 1);
    sem_init(&db, 0, 1);

    pthread_t readers[2], writers[2];

    // Create reader and writer threads
    for (int i = 0; i < 2; i++) {
        pthread_create(&readers[i], NULL, reader, (void *)(intptr_t)(i + 1));
        pthread_create(&writers[i], NULL, writer, (void *)(intptr_t)(i + 1));
    }

    // Wait for all threads to complete
    for (int i = 0; i < 2; i++) {
        pthread_join(readers[i], NULL);
        pthread_join(writers[i], NULL);
    }

    // Cleanup semaphores
    sem_destroy(&mutex);
    sem_destroy(&db);

    return 0;
}

void *reader(void *p) {
    printf("Reader %ld is trying to read\n", (intptr_t)p);
    sem_wait(&mutex); // Acquire mutex for reader count
    readercount++;
    if (readercount == 1) {
        sem_wait(&db); // First reader locks the database
    }
    sem_post(&mutex); // Release mutex for reader count

    printf("Reader %ld is reading\n", (intptr_t)p);
    sleep(1); // Simulate reading

    sem_wait(&mutex); // Acquire mutex for reader count
    readercount--;
    if (readercount == 0) {
        sem_post(&db); // Last reader unlocks the database
    }
    sem_post(&mutex); // Release mutex for reader count

    printf("Reader %ld finished reading\n", (intptr_t)p);
    return NULL;
}

void *writer(void *p) {
    printf("Writer %ld is trying to write\n", (intptr_t)p);
    sem_wait(&db); // Wait for database access

    printf("Writer %ld is writing\n", (intptr_t)p);
    sleep(1); // Simulate writing

    sem_post(&db); // Release database access
    printf("Writer %ld finished writing\n", (intptr_t)p);
    return NULL;
}

File: OS/code/readerwriter.py
import threading
import time

class ReaderWriter:
    def __init__(self):
        self.readers = 0
        self.lock = threading.Lock()
        self.writing = threading.Semaphore(1)

    def start_read(self):
        with self.lock:
            # Wait for a writer to finish
            print(f"Reader {threading.current_thread().name} is waiting to read.")
            self.readers += 1
            if self.readers == 1:
                self.writing.acquire()  # First reader locks the writer

        print(f"Reader {threading.current_thread().name} is reading.")
        time.sleep(1)  # Simulate reading

    def end_read(self):
        with self.lock:
            self.readers -= 1
            if self.readers == 0:
                self.writing.release()  # Last reader unlocks the writer

    def start_write(self):
        print(f"Writer {threading.current_thread().name} is waiting to write.")
        self.writing.acquire()  # Wait for readers to finish

        print(f"Writer {threading.current_thread().name} is writing.")
        time.sleep(1)  # Simulate writing

    def end_write(self):
        self.writing.release()  # Release the writer lock

def reader(rw):
    rw.start_read()
    rw.end_read()

def writer(rw):
    rw.start_write()
    rw.end_write()

if __name__ == "__main__":
    rw = ReaderWriter()

    # Creating readers and writers
    threads = []
    for _ in range(10):
        t = threading.Thread(target=reader, args=(rw,))
        threads.append(t)
        t.start()

    for _ in range(2):
        t = threading.Thread(target=writer, args=(rw,))
        threads.append(t)
        t.start()

    # Wait for all threads to complete
    for t in threads:
        t.join()
